transform/startf/ds/dataset.go

```diff
@@ -418,5 +464,95 @@ func (d *Dataset) AssignPreviousStructureEntries(entries int) {
 		// to worry that the structure is only partially filled.
 		d.ds.Structure = &dataset.Structure{}
 	}
-	d.ds.Structure.Entries = entries
+	d.ds.Structure.Entries = prev.Structure.Entries
+	return nil
+}
+
+func (d *Dataset) assignStructureFromDataframeColumns() error {
+	if d.bodyFrame == nil {
+		return nil
+	}
+	df, ok := d.bodyFrame.(*dataframe.DataFrame)
+	if !ok {
+		return fmt.Errorf("bodyFrame has invalid type %T", d.bodyFrame)
+	}
+
+	names, types := df.ColumnNamesTypes()
+	if names == nil || types == nil {
+		return nil
+	}
+
+	cols := make([]interface{}, len(names))
+	for i := range names {
+		cols[i] = map[string]string{
+			"title": names[i],
+			"type":  dataframeTypeToQriType(types[i]),
+		}
+	}
+
+	newSchema := map[string]interface{}{
+		"type": "array",
+		"items": map[string]interface{}{
+			"type":  "array",
+			"items": cols,
+		},
+	}
+	d.ds.Structure.Schema = newSchema
```

Comment:
need to defend against `Structure` being nil here. Same as in `AssignBodyFromDataframe`, but with no need for a shema b/c you're setting it:

```go
	st := d.ds.Structure
	if st == nil {
		st = &dataset.Structure{
			Format: "csv",
		}
	}
```